### PR TITLE
MGMT-8849: Avoid races when matching agents

### PR DESCRIPTION
**Pull request by avishayt**
To allow multiple instances of the provider to run simultaneously, we
now:
1. Update the Agent in one call, including a label referencing the
   AgentMachine.
2. Update the AgentMachine in one call.
3. If we did step 1 but failed on step2, use the label on the Agent to
   skip to step 2.

### CODE REVIEW

File: controllers/agentmachine_controller.go [hunk 1]
```diff
@@ -180,10 +173,44 @@ func (r *AgentMachineReconciler) handleDeletionFinalizer(ctx context.Context, lo
...
+		return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
+	}
+	if foundAgent != nil {
+		err = r.updateAgentMachineWithFoundAgent(ctx, log, agentMachine, foundAgent)
+		if err != nil {
+			log.WithError(err).Error("failed to update AgentMachine with found dangling agent")
+			return ctrl.Result{RequeueAfter: defaultRequeueAfterOnError}, err
+		} else {
+			return ctrl.Result{Requeue: true}, nil
+		}
```
Comment (eranco74) [inline controllers/agentmachine_controller.go:209]: I think you can refactor the function so this logic happens once (at the end).
It doesn't matter if the agent was dangling or not.

File: controllers/agentmachine_controller.go [hunk 2]
```diff
@@ -49,6 +49,7 @@ const (
 	defaultRequeueWaitingForAgentToBeInstalled = 20 * time.Second
 	defaultRequeueWaitingForAvailableAgent     = 30 * time.Second
 	AgentMachineFinalizerName                  = "agentmachine." + aiv1beta1.Group + "/deprovision"
+	DanglingAgentLabelKey                      = "agentMachineRef"
```
[the remark this answers is not in the document]
Comment (eranco74) [reply controllers/agentmachine_controller.go:52]: If I understand it correctly the label means the agent is `assigned to` / `referenced by` an AgentMachine.
 So `agentMachineRef` seems more accurate.

File: controllers/agentmachine_controller.go [hunk 3]
```diff
@@ -194,83 +221,141 @@ func (r *AgentMachineReconciler) findAgent(ctx context.Context, log logrus.Field
 	}
 
 	agents := &aiv1beta1.AgentList{}
-	if err := r.List(ctx, agents, &client.ListOptions{LabelSelector: selector}); err != nil {
+	if err = r.List(ctx, agents, &client.ListOptions{LabelSelector: selector}); err != nil {
```
Comment (eranco74) [inline controllers/agentmachine_controller.go:224]: I think the selector should filter out agents with `MachineAgentRef`

File: controllers/agentmachine_controller.go [hunk 4]
```diff
@@ -194,83 +221,141 @@ func (r *AgentMachineReconciler) findAgent(ctx context.Context, log logrus.Field
...
+				foundAgent = nil
+			} else {
+				break
+			}
 		}
 	}
 
 	if foundAgent == nil {
-		log.Info("Did not find any available Agents")
+		log.Info("Failed to own any available Agent")
```
[the remark this answers is not in the document]
Comment (avishayt) [reply controllers/agentmachine_controller.go:245]: Because we might have found agents but failed to update for some reason

**Review by eranco74 — COMMENTED** [review]
/lgtm

**Comment by osherdp** [comment]
/hold
Let's save resources, it does not seem to pass one of the CI jobs

**Comment by osherdp** [comment]
/hold
Let's save resources, it does not seem to pass...

**Pull request merged**